### PR TITLE
ternimal: init at 0.1.0-unstable-2017-12-31

### PR DESCRIPTION
**Pull request by Anomalocaridid**
## Description of changes

Adds [ternimal](https://github.com/p-e-w/ternimal), a program that renders an animated creature in a terminal.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by kirillrdy** [comment]
@Anomalocaridid I am not this project is a good fit to be added to nixpkgs

mainly because its just a few commits made 7 years ago

but if its packaged and it works, maybe you can still share the derivation on https://discourse.nixos.org/

**Comment by wegank** [comment]
I'm fine with adding the package as long as it works with the latest version of rustc.

(but I won't click the merge button until everyone agrees)

File: pkgs/by-name/te/ternimal/package.nix [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, rustc
+}:
+stdenv.mkDerivation rec {
+  pname = "ternimal";
+  version = "unstable-2017-12-31";
```
Comment (wegank) [inline pkgs/by-name/te/ternimal/package.nix:8]:
```suggestion
  version = "0.1.0-unstable-2017-12-31";
```

https://github.com/NixOS/nixpkgs/tree/master/pkgs#package-naming

File: pkgs/by-name/te/ternimal/package.nix [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+  src = fetchFromGitHub {
+    owner = "p-e-w";
+    repo = pname;
```
Comment (wegank) [inline pkgs/by-name/te/ternimal/package.nix:12]:
```suggestion
    repo = "ternimal";
```

Technically, it is unrelated to `pname`, which is rather arbitrary.

File: pkgs/by-name/te/ternimal/package.nix [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    rustc -O $src/ternimal.rs -o $out/bin/ternimal
+
+    runHook postBuild
+  '';
+
+  meta = {
+    description = "Simulate a lifeform in the terminal";
+    homepage = "https://github.com/p-e-w/ternimal";
+    license = lib.licenses.gpl3Only;
+    maintainers = with lib.maintainers; [ anomalocaris ];
```
Comment (wegank) [inline pkgs/by-name/te/ternimal/package.nix:32]: `meta.platforms` isn't set.

https://github.com/NixOS/nixpkgs/tree/master/pkgs#new-packages

File: pkgs/by-name/te/ternimal/package.nix (hunk 2 above)
Comment (kirillrdy) [reply pkgs/by-name/te/ternimal/package.nix:12]: more info here https://github.com/NixOS/nixpkgs/issues/277994

File: pkgs/by-name/te/ternimal/package.nix [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  src = fetchFromGitHub {
+    owner = "p-e-w";
+    repo = "ternimal";
+    rev = "e7953b4f80e514899e0920f0e36bb3141b685122";
+    hash = "sha256-uIxuwRvStvlC/YiolOvWZd45Qg3b86jsZQ000zZMo3M=";
+  };
+
+  nativeBuildInputs = [ rustc ];
+
+  buildPhase = ''
```
Comment (kirillrdy) [inline pkgs/by-name/te/ternimal/package.nix:19]: this should be split into buildPhase and installPhase, i know may seem like an overkill for something this simple

File: pkgs/by-name/te/ternimal/package.nix [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin
+    cp ternimal $out/bin
+
+    runHook postInstall
+  '';
+
+  meta = {
```
Comment (kirillrdy) [inline pkgs/by-name/te/ternimal/package.nix:36]: sorry for adhoc feedback, i should have spotted it earlier

`mainProgram` should be set 

https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#meta-attributes

Comment (Anomalocaridid) [reply pkgs/by-name/te/ternimal/package.nix:36]: No worries. It happens.

**Review by kirillrdy — COMMENTED** [review]
image

Result of `nixpkgs-review pr 301145` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ternimal</li>
  </ul>
</details>

**Comment by Anomalocaridid** [comment]
Is there anything else I need to do before this can be merged?

**Review by kirillrdy — APPROVED** [review]

**Comment by Anomalocaridid** [comment]
Thank you!

**Pull request merged**